message: port advanced rpc compression from scylla-enterprise.git to scylladb.git

michoecho opened the pull request:
This is a forward port (from scylla-enterprise) of additional compression options (zstd, dictionaries shared across messages) for inter-node network traffic. It works as follows:

After the patch, messaging_service (Scylla's interface for all inter-node communication)
compresses its network traffic with compressors managed by
the new advanced_rpc_compression::tracker. Those compressors compress with lz4,
but can also be configured to use zstd as long as a CPU usage limit isn't crossed.

A precomputed compression dictionary can be fed to the tracker. Each connection
handled by the tracker will then start a negotiation with the other end to switch
to this dictionary, and when it succeeds, the connection will start being compressed using that dictionary.

All traffic going through the tracker is passed as a single merged "stream" through dict_sampler.
dictionary_service has access to the dict_sampler.
On chosen nodes (in the "usual" configuration: the Raft leader), it uses the sampler to maintain
a random multi-megabyte sample of the sampler's stream. Every several minutes,
it copies the sample, trains a compression dictionary on it (by calling zstd's
training library via the alien_worker thread) and publishes the new dictionary
to system.dicts via Raft's write_mutation command.

This update triggers (eventually) a callback on all nodes, which feeds the new dictionary
to advanced_rpc_compression::tracker, and this switches (eventually) all inter-node connections
to this dictionary.

scylladb-promoter commented:
## Docs Preview :book:
Docs Preview for this pull request is available [here](http://previews.docs.scylladb.com/scylladb/22032)

**Changed Files:**
- [operating-scylla/procedures/config-change/advanced-internode-compression](http://previews.docs.scylladb.com/scylladb/22032/operating-scylla/procedures/config-change/advanced-internode-compression)

**Note:** This preview will be available for 30 days and will be automatically deleted after that period. You can manually trigger a new build by committing changes.

tchaikov commented:
```
/home/runner/work/scylladb/scylladb/docs/operating-scylla/procedures/config-change/advanced-internode-compression.rst: WARNING: document isn't included in any toctree
```

probably we failed to port some changes from 76f1ee6785ad997c4f3c4d58cfd827c712e0df1a

tchaikov commented on conf/scylla.yaml:
```diff
@@ -477,6 +477,47 @@ commitlog_total_space_in_mb: -1
+# by this node on compressing inter-node traffic with zstd.
+#
+# Depending on the workload, enabling zstd might have a drastic negative
+# effect on performace, so it shouldn't be done lightly.
```
could take this opportunity to fix the typo.

tchaikov commented on docs/dev/advanced_rpc_compression.md:
ditto. might want to fix the typo here.

tchaikov commented on docs/dev/advanced_rpc_compression.md:
ditto.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14049/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14049/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/advanced_rpc_compressor_test
:small_blue_diamond: boost/dict_trainer_test
:small_blue_diamond: boost/reservoir_sampling_test
:small_blue_diamond: boost/stream_compressor_test
:small_blue_diamond: topology_custom/test_rpc_compression
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/864/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2492/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13620/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5607/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/268/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14049/)
#### Build Details:
- Duration: 4 hr 33 min
- Builder: spider2.cloudius-systems.com

michoecho commented:
v2: Addressed Kefu's comments.
- Fixed 3 typos.
- Added `docs/operating-scylla/procedures/config-change/advanced-internode-compression.rst` to `docs/operating-scylla/procedures/config-change/index.rst`, same as in scylladb/scylla-enterprise@76f1ee6785ad997c4f3c4d58cfd827c712e0df1a

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14108/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14108/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/advanced_rpc_compressor_test
:small_blue_diamond: boost/dict_trainer_test
:small_blue_diamond: boost/reservoir_sampling_test
:small_blue_diamond: boost/stream_compressor_test
:small_blue_diamond: topology_custom/test_rpc_compression
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13662/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5649/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2535/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/907/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/311/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14108/)
#### Build Details:
- Duration: 9 hr 43 min
- Builder: i-04c0b8cff992bb02d (m5ad.8xlarge)

michoecho commented:
@avikivity Ping for review/merge.

avikivity commented:
Ugh the typo fixes are unfortunate, they will complicate the master->enterprise merge for me.

michoecho commented:
> Ugh the typo fixes are unfortunate, they will complicate the master->enterprise merge for me.

Shall I revert them?

avikivity commented:
> > Ugh the typo fixes are unfortunate, they will complicate the master->enterprise merge for me.
> 
> Shall I revert them?

No, I managed to resolve the problems.

I saw that scylladb/scylla-enterprise@e0800d1421031bc6540de6dfa50c6986678b11d4 missed the boat and has to be forward-ported separately.
